org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/hierarchy/AbstractCallHierarchyViewPart.java

```diff
@@ -0,0 +1,180 @@
+		if (callHierarchyBuilder != null) {
+			callHierarchyBuilder.setHierarchyType(hierarchyType);
+		}
+		refresh(new NullProgressMonitor());
+	}
+
+	@Override
+	public void refresh(IProgressMonitor monitor) {
+		CallHierarchyType hierarchyType = getCallHierarchyBuilder() != null ? getCallHierarchyBuilder().getHierarchyType() : null;
+		for (SetHierarchyTypeAction setHierarchyTypeAction : setHierarchyTypeActions) {
+			setHierarchyTypeAction.setChecked(hierarchyType == setHierarchyTypeAction.getHierarchyType());
+		}
+		super.refresh(monitor);
+	}
+
+	@Override
+	public void setRoot(IHierarchyRoot root) {
+		callHierarchyViewer.setInput(root);
+		setFocus();
+	}
+
+	@Override
+	public void createPartControl(Composite parent) {
+		Composite layout = createLayout(parent);
+		callHierarchyViewer = createCallHierarchyViewer(layout);
+		locationViewer = createLocationViewer(layout);
+		navigationService.installNavigationSupport(locationViewer);
+		navigationService.installNavigationSupport(callHierarchyViewer);
+		callHierarchyViewer.addSelectionChangedListener(this);
+		addActions(getViewSite().getActionBars().getToolBarManager());
+	}
+
+	protected void addActions(IToolBarManager toolBarManager) {
+		toolBarManager.add(refreshAction);
+		for (SetHierarchyTypeAction setHierarchyTypeAction : setHierarchyTypeActions) {
+			toolBarManager.add(setHierarchyTypeAction);
+		}
+	}
+
+	protected Composite createLayout(Composite parent) {
+		return new SashForm(parent, SWT.NONE);
+	}
+
+	protected TreeViewer createCallHierarchyViewer(Composite parent) {
+		TreeViewer treeViewer = new TreeViewer(parent);
+		treeViewer.getControl().setLayoutData(new GridData(GridData.FILL_BOTH));
+		treeViewer.setUseHashlookup(true);
+		treeViewer.setAutoExpandLevel(2);
+		treeViewer.setLabelProvider(createHierarchyLabelProvider());
+		treeViewer.setContentProvider(createHierarchyContentProvider());
+		return treeViewer;
+	}
+
+	protected IContentProvider createHierarchyContentProvider() {
+		return new HierarchyTreeContentProvider(this);
+	}
+
+	protected abstract IBaseLabelProvider createHierarchyLabelProvider();
+
+	protected TableViewer createLocationViewer(Composite parent) {
+		TableViewer locationViewer = new TableViewer(parent);
+		locationViewer.setContentProvider(new ArrayContentProvider());
+		locationViewer.setLabelProvider(createLocationLabelProvider());
+
+		TableLayout layout = new TableLayout();
+		locationViewer.getTable().setLayout(layout);
+		locationViewer.getTable().setHeaderVisible(true);
+
+		Pair<String, ColumnLayoutData>[] locationColumnDescriptions = getLocationColumnDescriptions();
+		IntStream.range(0, locationColumnDescriptions.length).forEach(index -> {
+			layout.addColumnData(locationColumnDescriptions[index].getValue());
+			createColumn(locationViewer.getTable(), locationColumnDescriptions[index], index);
+		});
+
+		return locationViewer;
+	}
+
+	protected Pair<String, ColumnLayoutData>[] getLocationColumnDescriptions() {
```

Comment:
@szarnekow Replacing `Pair` with some Java equivalent would break the downstream code, which is why I kept it as it is.